File: sdk/typescript/src/guard/custom.ts
/**
 * Custom guard -- user-defined regex or callable.
 *
 * Must stay in sync with: sdk/python/src/openlit/guard/custom.py
 */

import {
  Guard,
  GuardConfigError,
  GuardPhase,
  GuardResult,
  makeGuardResult,
  GuardOptions,
} from './base';

export interface CustomOptions extends GuardOptions {
  pattern?: string;
  callable?: (text: string) => GuardResult;
  phases?: string[];
}

export class Custom extends Guard {
  readonly name = 'custom';
  readonly phases: GuardPhase[];

  private readonly _pattern: RegExp | null;
  private readonly _callable?: (text: string) => GuardResult;

  constructor(opts: CustomOptions = {}) {
    super({ action: opts.action ?? 'deny', maxScanLength: opts.maxScanLength });

    if (!opts.pattern && !opts.callable) {
      throw new GuardConfigError("Custom guard needs at least 'pattern' or 'callable'");
    }

    this._pattern = opts.pattern ? new RegExp(opts.pattern) : null;
    this._callable = opts.callable;

    if (opts.phases) {
      this.phases = opts.phases.map((p) => p as GuardPhase);
    } else {
      this.phases = [GuardPhase.PREFLIGHT, GuardPhase.POSTFLIGHT];
    }
  }

  evaluate(text: string): GuardResult {
    if (this._pattern) {
      const match = this._pattern.exec(text);
      if (match) {
        return makeGuardResult({
          action: this._action,
          score: 1.0,
          guardName: this.name,
          classification: 'pattern_match',
          explanation: `Custom pattern matched: '${match[0]}'`,
        });
      }
    }

    if (this._callable) {
      return this._callable(text);
    }

    return makeGuardResult({ guardName: this.name });
  }
}

File: sdk/python/src/openlit/guard/custom.py
"""
Custom guard -- user-defined regex or callable.
"""

from __future__ import annotations

import re
from typing import Callable, List, Optional

from openlit.guard._base import Guard, GuardConfigError, GuardPhase, GuardResult


class Custom(Guard):
    """
    A user-defined guard backed by a regex pattern and/or a callable.

    Parameters
    ----------
    action : str
        ``"deny"``, ``"warn"``, or ``"redact"``.
    pattern : str, optional
        A regex pattern.  If it matches, the guard triggers.
    callable : callable, optional
        ``Callable[[str], GuardResult]`` for arbitrary logic.
    phases : list[str], optional
        ``["preflight"]``, ``["postflight"]``, or both (default).
    """

    name = "custom"

    def __init__(
        self,
        action: str = "deny",
        pattern: Optional[str] = None,
        callable: Optional[Callable[[str], GuardResult]] = None,  # pylint: disable=redefined-builtin
        phases: Optional[List[str]] = None,
        **kwargs,
    ):
        super().__init__(action=action, **kwargs)

        if pattern is None and callable is None:
            raise GuardConfigError(
                "Custom guard needs at least 'pattern' or 'callable'"
            )

        self._pattern: Optional[re.Pattern] = re.compile(pattern) if pattern else None
        self._callable = callable

        if phases:
            self.phases = tuple(GuardPhase(p) for p in phases)
        else:
            self.phases = (GuardPhase.PREFLIGHT, GuardPhase.POSTFLIGHT)

    def evaluate(self, text: str) -> GuardResult:
        if self._pattern:
            match = self._pattern.search(text)
            if match:
                return GuardResult(
                    action=self._action,
                    score=1.0,
                    guard_name=self.name,
                    classification="pattern_match",
                    explanation=f"Custom pattern matched: {match.group()!r}",
                )

        if self._callable is not None:
            return self._callable(text)

        return GuardResult(guard_name=self.name)
